### PR TITLE
[SPARK-43021][SQL] `CoalesceBucketsInJoin` not work when using AQE

### PR DESCRIPTION
**Pull request by zzzzming95**
### What changes were proposed in this pull request?

Add `CoalesceBucketsInJoin` to AQE `preprocessingRules`. 

### Why are the changes needed?

Previously optimized bucket join: 'CoalesceBucketsInJoin'` : https://github.com/apache/spark/pull/28123

But when using AQE , `CoalesceBucketsInJoin` can not match beacuse the top of the spark plan is `AdaptiveSparkPlan`.

The code : 
```
  val spark = SparkSession.builder()
    .appName("BucketJoin")
    .master("local[*]")
    .config("spark.sql.adaptive.enabled", true)
    .config("spark.driver.memory", "4")
    .config("spark.sql.autoBroadcastJoinThreshold", "-1")
    .config("spark.sql.bucketing.coalesceBucketsInJoin.enabled", true)
    .enableHiveSupport()
    .getOrCreate()

    val df1 = (0 until 20).map(i => (i % 5, i % 13, i.toString)).toDF("i", "j", "k")
    val df2 = (0 until 20).map(i => (i % 7, i % 11, i.toString)).toDF("i", "j", "k")
    df1.write.format("parquet").bucketBy(4, "i").saveAsTable("t1")
    df2.write.format("parquet").bucketBy(2, "i").saveAsTable("t2")
    val t1 = spark.table("t1")
    val t2 = spark.table("t2")
    val joined = t1.join(t2, t1("i") === t2("i"))
    joined.explain()
```

Before the PR
```
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- SortMergeJoin [i#50], [i#56], Inner
   :- Sort [i#50 ASC NULLS FIRST], false, 0
   :  +- Filter isnotnull(i#50)
   :     +- FileScan parquet spark_catalog.default.t1[i#50,j#51,k#52] Batched: true, Bucketed: true, DataFilters: [isnotnull(i#50)], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/Users/shezhiming/gh/zzzzming_new/spark/spark-warehouse/t1], PartitionFilters: [], PushedFilters: [IsNotNull(i)], ReadSchema: struct<i:int,j:int,k:string>, SelectedBucketsCount: 4 out of 4
   +- Sort [i#56 ASC NULLS FIRST], false, 0
      +- Exchange hashpartitioning(i#56, 4), ENSURE_REQUIREMENTS, [plan_id=78]
         +- Filter isnotnull(i#56)
            +- FileScan parquet spark_catalog.default.t2[i#56,j#57,k#58] Batched: true, Bucketed: false (disabled by query planner), DataFilters: [isnotnull(i#56)], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/Users/shezhiming/gh/zzzzming_new/spark/spark-warehouse/t2], PartitionFilters: [], PushedFilters: [IsNotNull(i)], ReadSchema: struct<i:int,j:int,k:string>

```

After the PR output:
```
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- SortMergeJoin [i#50], [i#56], Inner
   :- Sort [i#50 ASC NULLS FIRST], false, 0
   :  +- Filter isnotnull(i#50)
   :     +- FileScan parquet spark_catalog.default.t1[i#50,j#51,k#52] Batched: true, Bucketed: true, DataFilters: [isnotnull(i#50)], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/Users/shezhiming/gh/zzzzming_new/spark/spark-warehouse/t1], PartitionFilters: [], PushedFilters: [IsNotNull(i)], ReadSchema: struct<i:int,j:int,k:string>, SelectedBucketsCount: 4 out of 4 (Coalesced to 2)
   +- Sort [i#56 ASC NULLS FIRST], false, 0
      +- Filter isnotnull(i#56)
         +- FileScan parquet spark_catalog.default.t2[i#56,j#57,k#58] Batched: true, Bucketed: true, DataFilters: [isnotnull(i#56)], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/Users/shezhiming/gh/zzzzming_new/spark/spark-warehouse/t2], PartitionFilters: [], PushedFilters: [IsNotNull(i)], ReadSchema: struct<i:int,j:int,k:string>, SelectedBucketsCount: 2 out of 2

```

Additional Notes:

We don't add CoalesceBucketsInJoin to `AdaptiveSparkPlanExec#queryStageOptimizerRules` because queryStageOptimizerRules is not applied at the beginning of the init plan. Instead, they are applied in the createQueryStages() method. And createQueryStages() is bottom-up, which causes the exchange to be eliminated to be wrapped in a layer of ShuffleQueryStage first, making CoalesceBucketsInJoin unrecognizable.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

add UT

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
cc @imback82, @cloud-fan , @viirya , @sunchao

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/InsertAdaptiveSparkPlan.scala [hunk 1]
```diff
@@ -60,6 +61,7 @@ case class InsertAdaptiveSparkPlan(
           val subqueryMap = buildSubqueryMap(plan)
           val planSubqueriesRule = PlanAdaptiveSubqueries(subqueryMap)
           val preprocessingRules = Seq(
+            CoalesceBucketsInJoin,
             planSubqueriesRule)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/InsertAdaptiveSparkPlan.scala:65]: Hmm, why don't put into `queryStagePreparationRules`? These no-op physical rules including `DisableUnnecessaryBucketedScan` are put there.

**Review by viirya — COMMENTED** [review]
We may need adding test case.

**Comment by dongjoon-hyun** [comment]
Maybe, no? If this is not working properly before, we cannot enable this configuration at Apache Spark 3.5.0. Since we need to wait for one release cycle, we may be able to do that at Apache Spark 3.6.0 if we want.
> One more question , it time to make the default value of `SQLConf.COALESCE_BUCKETS_IN_JOIN_ENABLED` as true ?

**Comment by dongjoon-hyun** [comment]
Please rebase to the `master` branch once more, @zzzzming95 .
> The CI build failure doesn't seem to be caused by this patch, can you take a look?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala [hunk 2]
```diff
@@ -118,6 +118,7 @@ case class AdaptiveSparkPlanExec(
     val ensureRequirements =
       EnsureRequirements(requiredDistribution.isDefined, requiredDistribution)
     Seq(
+      CoalesceBucketsInJoin,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:121]: shall we put it in `queryStageOptimizerRules`?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:121]: rules in `queryStageOptimizerRules` are invoked less often which is more efficient. The rule `CoalesceBucketsInJoin` does not change plan partitioning and seems can be put in `queryStageOptimizerRules`

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:121]: Can we spend a bit of time understanding why? Then we can write a code comment to explain it and following developers won't try to move this rule to `queryStageOptimizerRules` ever.

**Comment by dongjoon-hyun** [comment]
To be clear, this PR didn't get any approval yet, @zzzzming95 .
> Please merge to master . Thanks ~

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala (hunk 2 above)
Comment (wangyum) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:121]: `CoalesceBucketsInJoin` should before `EnsureRequirements`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala [hunk 3]
```diff
@@ -118,6 +118,7 @@ case class AdaptiveSparkPlanExec(
     val ensureRequirements =
       EnsureRequirements(requiredDistribution.isDefined, requiredDistribution)
     Seq(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:120]: let's add comments after this line:
```
// CoalesceBucketsInJoin can help eliminate shuffles and must be run before
// EnsureRequirements
```

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala [hunk 4]
```diff
@@ -1020,21 +1019,25 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils with Adapti
...
-
-          val scans = plan.collect {
-            case f: FileSourceScanExec if f.optionalNumCoalescedBuckets.isDefined => f
-          }
-          if (expectedCoalescedNumBuckets.isDefined) {
-            assert(scans.length == 1)
-            assert(scans.head.optionalNumCoalescedBuckets == expectedCoalescedNumBuckets)
-          } else {
-            assert(scans.isEmpty)
+                    query: String,
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala:1022]: nit: the indentation is wrong now, can we restore to 4 spaces as before?

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala:1022]: Can we follow https://github.com/apache/spark/pull/40731/files#diff-1dd0d5a38f73f2993e5852f759a3934396c083d4fc4cc334e73ccc8eb929a717R1013 to update the test?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala [hunk 5]
```diff
@@ -1020,21 +1019,25 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils with Adapti
...
+                    expectedNumShuffles: Int,
+                    expectedCoalescedNumBuckets: Option[Int]): Unit = {
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala:1024]:
```suggestion
            query: String,
            expectedNumShuffles: Int,
            expectedCoalescedNumBuckets: Option[Int]): Unit = {
```

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by neshkeev** [comment]
I was the OP of the issue in the jira.

Thank you for the fix, but I discovered a weird behavior when hints applied and I don't know how to interpret it. Please check [SPARK-43326](https://issues.apache.org/jira/browse/SPARK-43326) I filled

**Pull request closed without merging**